File: backend/src/SmartHome.Api/Devices/ControlDeviceRequest.cs
using System.ComponentModel.DataAnnotations;
using SmartHome.Domain.Devices.Light;
using SmartHome.Domain.Devices.Thermostat;
using SmartHome.Domain.Devices.Fan;
using SmartHome.Domain.Commands;

namespace SmartHome.Api.Devices;

/// <summary>
/// DTO used to represent a command request from the client.
/// </summary>

public class ControlDeviceRequest
{
    public DeviceCommandType Command { get; set; }
    public int? Brightness { get; set; }
    public string? Color { get; set; }
    public FanSpeed? FanSpeed { get; set; }
    public ThermostatMode? ThermostatMode { get; set; }
    public int? TargetTemperature { get; set; }

}

File: frontend/smart-home-ui/src/app/devicemodels/controldevicerequest.ts
import { FanSpeed } from '../types/fanspeed';
import { ThermostatMode } from '../types/thermostatmode';
import { DeviceCommandType } from '../types/devicecommandtype';

/** Request body for sending a command to a device. */
export interface ControlDeviceRequest {
  command: DeviceCommandType;

  brightness?: number | null;

  color?: string | null;

  fanSpeed?: FanSpeed | null;

  thermostatMode?: ThermostatMode | null;

  targetTemperature?: number | null;

  isLocked?: boolean | null;
}
